folding-schemes/src/folding/hypernova/decider_eth.rs

```diff
@@ -0,0 +1,309 @@
+        Ok((pp, vp))
+    }
+
+    fn prove(
+        mut rng: impl RngCore + CryptoRng,
+        pp: Self::ProverParam,
+        folding_scheme: FS,
+    ) -> Result<Self::Proof, Error> {
+        let (snark_pk, cs_pk): (S::ProvingKey, CS1::ProverParams) = pp;
+
+        let circuit = DeciderEthCircuit::<C1, GC1, C2, GC2, CS1, CS2>::from_hypernova::<FC, MU, NU>(
+            folding_scheme.into(),
+        )?;
+
+        let snark_proof = S::prove(&snark_pk, circuit.clone(), &mut rng)
+            .map_err(|e| Error::Other(e.to_string()))?;
+
+        let rho_Fr = circuit.rho.unwrap();
+        let W_i1 = circuit.W_i1.unwrap();
+
+        // get the challenges that have been already computed when preparing the circuit inputs in
+        // the above `from_hypernova` call
+        let challenge_W = circuit
+            .kzg_challenge
+            .ok_or(Error::MissingValue("kzg_c_W".to_string()))?;
+
+        // generate KZG proofs
+        let U_cmW_proof = CS1::prove_with_challenge(
+            &cs_pk,
+            challenge_W,
+            &W_i1.w,
+            &C1::ScalarField::zero(),
+            None,
+        )?;
+
+        Ok(Self::Proof {
+            snark_proof,
+            kzg_proof: U_cmW_proof,
+            rho: rho_Fr,
+            U_i1: circuit.U_i1.unwrap(),
+            kzg_challenge: challenge_W,
+        })
+    }
+
+    fn verify(
+        vp: Self::VerifierParam,
+        i: C1::ScalarField,
+        z_0: Vec<C1::ScalarField>,
+        z_i: Vec<C1::ScalarField>,
+        // we don't use the instances at the verifier level, since we check them in-circuit
+        _running_instance: &Self::CommittedInstance,
+        _incoming_instance: &Self::CommittedInstance,
+        proof: &Self::Proof,
+    ) -> Result<bool, Error> {
+        if i <= C1::ScalarField::one() {
+            return Err(Error::NotEnoughSteps);
+        }
+
+        let (pp_hash, snark_vk, cs_vk): (C1::ScalarField, S::VerifyingKey, CS1::VerifierParams) =
+            vp;
+
+        // Note: the NIMFS proof is checked inside the DeciderEthCircuit, which ensures that the
+        // 'proof.U_i1' is correctly computed
+
+        let (cmC_x, cmC_y) = NonNativeAffineVar::inputize(proof.U_i1.C)?;
+
+        let public_input: Vec<C1::ScalarField> = [
+            vec![pp_hash, i],
+            z_0,
+            z_i,
+            // U_i+1:
+            cmC_x,
+            cmC_y,
+            vec![proof.U_i1.u],
+            proof.U_i1.x.clone(),
+            proof.U_i1.r_x.clone(),
+            proof.U_i1.v.clone(),
+            vec![proof.kzg_challenge, proof.kzg_proof.eval, proof.rho],
+        ]
+        .concat();
+
+        let snark_v = S::verify(&snark_vk, &public_input, &proof.snark_proof)
+            .map_err(|e| Error::Other(e.to_string()))?;
+        if !snark_v {
+            return Err(Error::SNARKVerificationFail);
+        }
+
+        // we're at the Ethereum EVM case, so the CS1 is KZG commitments
```

Comment:
I don't understand the comment. We are in EVM case in the Solidity env. But here you could definitely use Pedersen too. 

Do I understand correctly? 
I mean, the circuit verified in the EVM makes sense with KZG ofc. But within this fn (which is not executed within the EVM) it's not clear that only KZG is valid. 
Although ofc, the design of the entire circuit is based on the fact that we will do a KZG check not a pedersen one for the verification.

Idk, while correct. The comment is a bit confusing IMO.

Reply:
this is the specially designed Deicder to be verified in Ethereum's EVM (file [`hypernova/decider_eth.rs`](https://github.com/privacy-scaling-explorations/sonobe/blob/hypernova-decider-eth/folding-schemes/src/folding/hypernova/decider_eth.rs)), if you use Pedersen at the G1 points, then you can not go for the verification in the EVM, and the circuit of this specific Decider will not fit, neither the decider logic itself.

Reply:
Sure. I just got confused with my own thoughts I assume hahaha